akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -41,112 +54,166 @@ class AffinityPool(numThreads: Int, affinityGroupSize: Int, tf: ThreadFactory) e
 
   private def tryEnqueue(command: Runnable) = {
     // does a mod on the hash code to determine which queue to put task in
-    val queueIndex = Math.abs(command.hashCode()) & (numThreads - 1)
+    val queueIndex = Math.abs(command.hashCode()) & (parallelism - 1)
     workQueues(queueIndex).add(command)
   }
 
-  private def onWorkerExit(w: ThreadPoolWorker): Unit = {
-    locked(mainLock) {
+  /**
+   * Each worker should go through that method while terminating.
+   * In turn each worker is responsible for modifying the pool
+   * state accordingly. For example if this is the last worker
+   * and the queue is empty and we are in a ShuttingDown state
+   * the worker can transition the pool to ShutDown and attempt
+   * termination
+   *
+   * Furthermore, if this worker has experienced abrupt termination
+   * due to an exception being thrown in user code, the worker is
+   * responsible for adding one more worker to compensate for its
+   * own termination
+   *
+   */
+  private def onWorkerExit(w: ThreadPoolWorker, abruptTermination: Boolean): Unit =
+    locked(bookKeepingLock) {
       workers.remove(w)
+      if (workers.isEmpty && !abruptTermination && poolState >= ShuttingDown) {
+        poolState = ShutDown // transition to shutdown and try to transition to termination
+        attemptPoolTermination()
+      }
+      if (abruptTermination && poolState == Running)
+        addWorker(workers, w.q)
     }
+
+  override def execute(command: Runnable): Unit = {
+    if (command == null)
+      throw new NullPointerException
+    if (!(poolState == Running && tryEnqueue(command)))
+      reject(command)
   }
 
-  private def onWorkerFailure(w: ThreadPoolWorker): Unit = {
-    locked(mainLock) {
-      workers.remove(w)
-      addWorker(workers, w.q)
+  private def reject(command: Runnable) = throw new RejectedExecutionException("Task " + command.toString + " rejected from " + this.toString)
+
+  override def awaitTermination(timeout: Long, unit: TimeUnit): Boolean = {
+
+    // recurse until pool is terminated or time out reached
+    @tailrec
+    def awaitTermination(nanos: Long): Boolean = poolState match {
+      case Terminated ⇒ true
+      case _ ⇒
+        if (nanos <= 0) false
+        else
+          awaitTermination(terminationCondition.awaitNanos(nanos))
+    }
```

Comment:
I think it would be tidier with something like:

```scala
if (poolState == Terminated) true
else if (nanos <= 0) false
else awaitTermination(terminationCondition.awaitNanos(nanos))
```